calling-apis: content (#15)
Ada-Developers-Academy/core-unit-3: calling-apis/intro-to-asynchronous-programming.md

# Intro to Asynchronous Programming

## Learning Goals

- Define asynchronous programming
- Compare synchronous and asynchronous programming

## Introduction: A Pizza Metaphor

It's Friday, and Mica is exhausted and spending the evening at home. It's dinner time, and they have no food in their freezer nor fridge. They aren't going to go grocery shopping, and nobody will make dinner for them tonight.

Instead, they'll make a phone call to their favorite restaurant (who's not on an app) and order dinner for themself for delivery.

There's a series of things that happen between the moment Mica makes the order all the way to the point where they take the first bite.

> In pain-staking detail, let's outline all the actions Mica would do, beginning with picking up the phone, and ending with eating the meal.

Reviewer: no hyphenation (painstaking derives from "taking pains" rather than "staking pain")
```suggestion
In painstaking detail, let's outline all the actions Mica would do, beginning with picking up the phone, and ending with eating the meal.
```

> In pain-staking detail, let's outline all the actions Mica would do, beginning with picking up the phone, and ending with eating the meal.

Reviewer: no hyphenation (painstaking derives from "taking pains" rather than "staking pain")

```suggestion
In painstaking detail, let's outline all the actions Mica would do, beginning with picking up the phone, and ending with eating the meal.
```

> <br/>

Reviewer 2: Feel free to ignore I just don't like unneeded HTML.

```suggestion
```

Reviewer: Without this br, learn crams the detail section up against the previous paragraph (even though there are new lines!)

<details style="max-width: 700px; margin: auto;">
  <summary>
    No, seriously. Take one minute to make a list of the actions Mica would do. After that, then read more here.
  </summary>

We may imagine that what Mica does is make the order, then wait for the order, and then eat.

It's actually way more likely that Mica will make the order, then **before the order arrives, they do other things,** like read, shower, sleep, watch TV. Mica is able to eat their food only after the food order has been received, cooked, assembled, assigned a driver, driven over, and delivered.

The food order was received, made, and delivered. However, while that was happening, Mica **did not wait for the food order to finish** in order to continue other actions. While waiting for the food, Mica didn't stop everything they were doing, sit on the couch, and stare at the wall. (Or maybe they did!)

> <br/>

Reviewer 2: Feel free to ignore, I just don't like unneeded HTML.

```suggestion
```

Reviewer: +1

This one results in a strangely tall gap between these two paragraphs.

1. When does this line of code "finish"?
1. What do we do when this line of code "finishes"?

</details>

## Vocabulary and Synonyms

| Vocab                    | Definition                                                                                                                                        | Synonyms          | How to Use in a Sentence                                                                                                                            |
| ------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |
| Flow of control          | The order of execution of programming statements.                                                                                                 | -                 | "When we trace code as it runs, we are following the flow of control"                                                                               |

> | Asynchronous programming | A method of programming that intentionally executes actions and processes outside of program flow, so that the program doesn't stop for a result. | Async programming | "Making API calls and loading pages on the web is usually asynchronous, and other code will execute without waiting for the result of an API call." |

Reviewer: even with async code, there is a program flow. It just isn't necessarily the same order as the written lines of code.
stop alone sounds like the program could be terminating. add wait to stress that it is... waiting
```suggestion
| Asynchronous programming | A method of programming that intentionally executes actions and processes outside of the typical program flow, so that the program doesn't stop and wait for a result. | Async programming | "Making API calls and loading pages on the web is usually asynchronous, and other code will execute without waiting for the result of an API call." |
```

We have been writing synchronous code this whole time!

For example, observe this Python code and this JavaScript code which do the same thing:

```python
def say_apples():
    print("apples")

def say_oranges():
    print("oranges")

say_apples()
say_oranges()
```

```js
const sayApples = () => {
  console.log("apples");
};

const sayOranges = () => {
  console.log("oranges");
};

sayApples();
sayOranges();
```

Both code snippets print out `apples` first, and then `oranges` second, because we call the function that prints `apples` first, and _then_ we call the function that prints `oranges` second. If we need to change the order of how these print, we need to change the order of the lines of code.

### Pizza Metaphor: Synchronous Code

Unlike the food-delivery example from above, synchronous code _executes and finishes_ in a very specific order.

Synchronous code may be more like calling the restaurant and making the order:

1. The restaurant asks what dishes you want
1. They wait for your answer
1. Then they proceed to the next question

## Asynchronous Programming

> [Asynchronous Programming](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Concepts) is a method of programming that intentionally breaks program flow in order to:

Reviewer: similar changes as in the vocab
bring the purpose (avoiding the stopping) up from the methodology
the bullets are actually describing how async achieves the goal of not stopping, so modify the lead-in line to reflect this.
```suggestion
[Asynchronous Programming](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Concepts) is a method of programming that intentionally breaks the typical flow of control in order to avoid having to stop and wait for a result.

Asynchronous programming accomplishes this by:
```

> 1. call actions/processes outside of program flow, so that the program doesn't stop for a result
> 2. define what happens when the result comes back

Reviewer: rephrase to fit the modified lead-in
```suggestion
1. calling actions/processes outside the usual program flow
2. defining what happens when the result eventually comes back
```

Asynchronous programming allows us to program behavior where some code _does not wait_ for other code to finish.

> To set up asynchronous programming, we will need to consider initially:

Reviewer: ```suggestion
To set up asynchronous programming, we will need to consider:
```

- Which specific technologies (tools and environment) to use that support running asynchronous code

> - While or after an asynchronous call, what should happen, and all the different cases of success and failure

Reviewer: while doesn't fit here (self-check by removing "or after" which leaves the ungrammatical "while an asynchronous call")
collapse the thinking about success/failure into "what should happen?"
```suggestion
- During or after an asynchronous call, what should happen for all the different cases of success and failure
```

| What do we do after the asynchronous call finishes?                                                                                 | What do we do after we receive the food delivery?                                                                                                                                                 |
| What do we do if the async call doesn't finish successfully?                                                                        | What do we do if the food never arrives? What do we do if the restaurant calls back and says that they are out of food? What do we do if the restaurant calls back and says they lost your order? |
| How can we ensure that every other part of the program runs correctly, without bugs, even if the asynchronous call hasn't finished? | How do we make sure that Mica isn't too hungry and orders another dinner? If Mica's friend joins them and they are hungry and wants to eat, what does Mica do?                                    |

## Summary

> Typically, we write a lot of code that runs in a very specific order (synchronous). However, certain technologies allow us to write asynchronous code, and we'll see plenty in JavaScript.

Reviewer: restate plenty of what. the technologies reference goes by quickly so I was left needing to go back and try to figure out what plenty referred to.
```suggestion
Typically, we write a lot of code that runs in a very specific order (synchronous). However, certain technologies allow us to write asynchronous code, and we'll see plenty of these technologies in JavaScript.
```
